polygeist/lib/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.cpp

```diff
@@ -13,45 +13,106 @@
+  if (def.isOperation())
+    os << *def.getOperation();
+  if (def.isInitialDefinition())
+    os << *def.getInitialDefinition();
+  return os;
+}
+
+bool Definition::operator==(const Definition &other) const {
+  if (isOperation() && other.isOperation())
+    return getOperation() == other.getOperation();
+  if (isInitialDefinition() && other.isInitialDefinition())
+    return true;
+  return false;
+}
+
+bool Definition::operator<(const Definition &other) const {
+  if (isOperation() && other.isOperation())
+    return getOperation() < other.getOperation();
+  if (isInitialDefinition() && other.isInitialDefinition())
+    return false; // InitialDefinition is a singleton.
+  return isInitialDefinition();
+}
+
 //===----------------------------------------------------------------------===//
 // ReachingDefinition
 //===----------------------------------------------------------------------===//
 
 raw_ostream &operator<<(raw_ostream &os, const ReachingDefinition &lastDef) {
   if (lastDef.valueToModifiers.empty() &&
       lastDef.valueToPotentialModifiers.empty())
-    return os.indent(4) << "<empty>\n";
-
-  auto printMap =
-      [&os](const DenseMap<Value, ReachingDefinition::ModifiersTy> &map,
-            StringRef title) {
-        for (const auto &entry : map) {
-          Value val = entry.first;
-          const ReachingDefinition::ModifiersTy &valModifiers = entry.second;
-
-          os.indent(4) << val << "\n";
-          os.indent(4) << title << "\n";
-          if (valModifiers.empty())
-            os.indent(6) << "<none>\n";
-          else {
-            for (Operation *op : valModifiers)
-              os.indent(6) << *op << "\n";
-          }
-          os << "\n";
-        }
-      };
+    return os.indent(4) << "<empty>";
+
+  using ModifiersTy = ReachingDefinition::ModifiersTy;
+  auto printMap = [&os](const DenseMap<Value, ModifiersTy> &map,
+                        StringRef title) {
+    for (const auto &entry : map) {
+      Value val = entry.first;
+      const ModifiersTy &valModifiers = entry.second;
+
+      os.indent(4) << val << "\n";
+      os.indent(4) << title << "\n";
+      if (valModifiers.empty())
+        os.indent(6) << "<none>\n";
+      else {
+        for (const std::shared_ptr<Definition> &def : valModifiers)
+          os.indent(6) << *def << "\n";
+      }
+    }
+  };
 
   printMap(lastDef.valueToModifiers, "mods:");
   printMap(lastDef.valueToPotentialModifiers, "pMods:");
   return os;
 }
 
+ReachingDefinition::ReachingDefinition(ProgramPoint p)
+    : AbstractDenseLattice(p) {
+  // Upon creating a new reaching definition at the start of a function, each
+  // memory argument is set to have an unknown initial value.
+  if (auto *block = p.dyn_cast<Block *>()) {
+    if (!block->isEntryBlock())
+      return;
+    if (auto funcOp = dyn_cast<FunctionOpInterface>(block->getParentOp())) {
+      for (Value arg : funcOp.getArguments()) {
+        if (isa<MemRefType>(arg.getType()))
```

Comment:
This adds a limitation to the analysis that it only works for memref, but I cannot think of an alternative.